apps/antalmanac/src/actions/ActionTypesStore.ts

```diff
@@ -0,0 +1,151 @@
+}
+
+export interface DeleteCourseAction extends BaseAction {
+    type: 'deleteCourse';
+    sectionCode: string;
+    term: string;
+}
+
+export interface AddCustomEventAction extends BaseAction {
+    type: 'addCustomEvent';
+    customEvent: RepeatingCustomEvent;
+    scheduleIndices: number[];
+}
+
+export interface DeleteCustomEventAction extends BaseAction {
+    type: 'deleteCustomEvent';
+    customEventId: number;
+}
+
+export interface EditCustomEventAction extends BaseAction {
+    type: 'editCustomEvent';
+    editedCustomEvent: RepeatingCustomEvent;
+    newScheduleIndices: number[];
+}
+
+export interface ChangeCustomEventColorAction extends BaseAction {
+    type: 'changeCustomEventColor';
+    customEventId: number;
+    newColor: string;
+}
+
+export interface ClearScheduleAction extends BaseAction {
+    type: 'clearSchedule';
+}
+
+export interface CopyScheduleAction extends BaseAction {
+    type: 'copySchedule';
+    to: number;
+}
+
+export interface ChangeCourseColorAction extends BaseAction {
+    type: 'changeCourseColor';
+    sectionCode: string;
+    term: string;
+    newColor: string;
+}
+
+export type ActionType =
+    | AddCourseAction
+    | DeleteCourseAction
+    | AddCustomEventAction
+    | DeleteCustomEventAction
+    | EditCustomEventAction
+    | ChangeCustomEventColorAction
+    | ClearScheduleAction
+    | CopyScheduleAction
+    | ChangeCourseColorAction
+    | UndoAction;
+
+class ActionTypesStore extends EventEmitter {
+    constructor() {
+        super();
+    }
+
+    async autoSaveSchedule(action: ActionType) {
+        const autoSave = typeof Storage !== 'undefined' && window.localStorage.getItem('autoSave') == 'true';
+        if (autoSave) {
+            const savedUserID = window.localStorage.getItem('userID');
+
+            if (savedUserID) {
+                this.emit('autoSaveStart');
+                await saveSchedule(savedUserID, true, true);
+                AppStore.unsavedChanges = false;
+                this.emit('autoSaveEnd');
+            }
+        } else {
+            const unsavedActionsString = window.localStorage.getItem('unsavedActions');
+            if (unsavedActionsString == null) {
+                const unsavedActions = [action];
+                localStorage.setItem('unsavedActions', JSON.stringify(unsavedActions));
+            } else {
+                const unsavedActions = JSON.parse(unsavedActionsString);
+                unsavedActions.push(action);
+                localStorage.setItem('unsavedActions', JSON.stringify(unsavedActions));
+            }
+        }
+    }
+
+    async loadScheduleFromLocalSave() {
+        const unsavedActionsString = window.localStorage.getItem('unsavedActions');
+        if (unsavedActionsString !== null) {
+            if (confirm('You have unsaved changes. Would you like to load them?')) {
+                for (const action of JSON.parse(unsavedActionsString)) {
+                    switch (action.type) {
```

Comment:
Inverse the conditional logic to utilize guard statements and reduce nesting. Also, use a type-safe function for accessing the raw string data, so that way you can easily factor in schema validation later if that's desired.

```tsx

// Somewhere else in the file
/**
  * (Somewhere else in the file)
  * This isn't type-safe within the function, 
  * but the parent scope using the function will have defined types.
  */
function parseUnsavedActionsString(unsavedActionsString: string): ActionType[] {
  try {
    return JSON.parse(unsavedActionsString)
  } catch {
    return []
  }
}

if (unsavedActionsString == null) return

if (!confirm('You have unsaved changes...')) return

const actions = parseActionsString(unsavedActionsString)

actions.forEach(action => {
  switch (action.type) {
    // ...
  }
})
```